File: native-android/app/src/main/java/site/smap/gugudan/features/session/ResultView.kt
package site.smap.gugudan.features.session

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.verticalScroll
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.AutoAwesome
import androidx.compose.material.icons.filled.Check
import androidx.compose.material.icons.filled.Close
import androidx.compose.material.icons.filled.EmojiEvents
import androidx.compose.material.icons.filled.LocalFireDepartment
import androidx.compose.material.icons.filled.Refresh
import androidx.compose.material.icons.filled.Speed
import androidx.compose.material3.Icon
import androidx.compose.material3.Text
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import kotlinx.coroutines.delay
import site.smap.gugudan.core.ModeKind
import site.smap.gugudan.core.Modes
import site.smap.gugudan.designsystem.*
import site.smap.gugudan.store.LocalGame

// 세션 결과 화면 (iOS ResultView.swift 이식)

@Composable
fun ResultView(engine: SessionEngine, done: SessionDone, onClose: () -> Unit) {
    val gg = LocalGG.current
    val game = LocalGame.current

    val result = done.result
    val commit = done.commit
    val def = Modes.def(result.mode)
    val total = result.answers.size
    val correct = total - done.wrongCount
    val accuracy = if (total > 0) Math.round(correct * 100.0 / total).toInt() else 0
    val avgMs = if (total > 0) Math.round(result.answers.sumOf { it.ms }.toDouble() / total).toInt() else 0
    val best = commit.score?.let { maxOf(game.state.bestScores[result.mode] ?: 0, it) }

    val headline = when {
        def.scored -> if (commit.isNewBest) "신기록 달성! 🏆" else "수고했어요!"
        accuracy >= 95 -> "완벽해요! 🎯"
        accuracy >= 70 -> "잘했어요!"
        else -> "좋은 시도예요!"
    }

    var confetti by remember { mutableStateOf(0) }
    var levelUp by remember { mutableStateOf(false) }

    LaunchedEffect(Unit) {
        if (accuracy >= 80 || commit.isNewBest) confetti += 1
        if (commit.leveledUp) { delay(600); levelUp = true }
    }

    Box(Modifier.fillMaxSize()) {
        Column(
            Modifier.fillMaxSize().verticalScroll(rememberScrollState())
                .statusBarsPadding().navigationBarsPadding()
                .padding(horizontal = 24.dp).padding(top = 24.dp, bottom = 32.dp),
            verticalArrangement = Arrangement.spacedBy(16.dp),
        ) {
            Column(verticalArrangement = Arrangement.spacedBy(4.dp)) {
                Text("${def.name} 완료", style = suite(FontWeight.Bold, 14), color = gg.textMuted)
                Text(headline, style = suite(FontWeight.ExtraBold, 28), color = gg.text)
            }

            // 점수형 모드 히어로
            if (def.scored && commit.score != null) {
                Column(
                    Modifier.fillMaxWidth().ggCard().padding(24.dp),
                    horizontalAlignment = Alignment.CenterHorizontally,
                    verticalArrangement = Arrangement.spacedBy(8.dp),
                ) {
                    if (commit.isNewBest) {
                        Pill(bg = gg.warning.copy(alpha = 0.15f), fg = gg.warning) {
                            Icon(Icons.Filled.EmojiEvents, null, Modifier.size(14.dp))
                            Text("신기록!")
                        }
                    }
                    Row(verticalAlignment = Alignment.Bottom) {
                        Text("${commit.score}", style = suite(FontWeight.ExtraBold, 56), color = gg.text)
                        Text("점", style = suite(FontWeight.Bold, 20), color = gg.textMuted,
                            modifier = Modifier.padding(bottom = 10.dp, start = 4.dp))
                    }
                    best?.let { Text("최고 기록 ${it}점", style = suite(FontWeight.Bold, 14), color = gg.textMuted) }
                }
            }

            // 핵심 수치 2×2
            Column(verticalArrangement = Arrangement.spacedBy(12.dp)) {
                Row(horizontalArrangement = Arrangement.spacedBy(12.dp)) {
                    StatCard(Modifier.weight(1f), Icons.Filled.Check, gg.success, "정확도", "$accuracy%")
                    StatCard(Modifier.weight(1f), Icons.Filled.Speed, gg.accent, "평균 속도", "%.1f초".format(avgMs / 1000.0))
                }
                Row(horizontalArrangement = Arrangement.spacedBy(12.dp)) {
                    StatCard(Modifier.weight(1f), Icons.Filled.AutoAwesome, gg.warning, "획득 XP", "+${commit.xpEarned}")
                    StatCard(Modifier.weight(1f), Icons.Filled.LocalFireDepartment, gg.danger, "최고 콤보", "${result.maxCombo}")
                }
            }

            // 마스터리 (단 집중)
            commit.table?.let { table ->
                Row(
                    Modifier.fillMaxWidth().ggCard(16.dp).padding(horizontal = 20.dp, vertical = 16.dp),
                    verticalAlignment = Alignment.CenterVertically,
                ) {
                    Column(Modifier.weight(1f)) {
                        Text("${table}단 마스터리", style = suite(FontWeight.Bold, 14), color = gg.text)
                        if (commit.improvedStars) {
                            Text("새 기록 달성!", style = suite(FontWeight.Bold, 12), color = gg.accent)
                        }
                    }
                    StarsView(commit.newStars, size = 24.dp)
                }
            }

            if (commit.goalReached) {
                Text(
                    "🎉 오늘의 목표를 달성했어요!",
                    style = suite(FontWeight.Bold, 14), color = gg.success, textAlign = TextAlign.Center,
                    modifier = Modifier.fillMaxWidth().clip(RoundedCornerShape(16.dp))
                        .background(gg.success.copy(alpha = 0.15f)).padding(vertical = 12.dp),
                )
            }

            Spacer(Modifier.weight(1f, fill = false))

            // 액션
            Column(verticalArrangement = Arrangement.spacedBy(10.dp)) {
                if (def.kind == ModeKind.FIXED && done.wrongCount > 0) {
                    GGButton(variant = GGButtonVariant.SURFACE, size = GGButtonSize.LG,
                        onClick = { engine.restart(retryWrong = true) }) {
                        Icon(Icons.Filled.Refresh, null, Modifier.size(20.dp))
                        Text("틀린 문제만 다시 (${done.wrongCount})")
                    }
                }
                GGButton(variant = GGButtonVariant.PRIMARY, size = GGButtonSize.LG,
                    onClick = { engine.restart(retryWrong = false) }) {
                    Text("한 판 더")
                }
                GGButton(variant = GGButtonVariant.GHOST, size = GGButtonSize.MD, onClick = onClose) {
                    Icon(Icons.Filled.Close, null, Modifier.size(16.dp))
                    Text("닫기")
                }
            }
        }

        ConfettiView(confetti)
        AchievementToast(commit.unlocked,
            Modifier.statusBarsPadding().padding(horizontal = 20.dp).padding(top = 12.dp))
        LevelUpOverlay(levelUp, commit.newLevel) { levelUp = false }
    }
}

@Composable
private fun StatCard(
    modifier: Modifier,
    icon: androidx.compose.ui.graphics.vector.ImageVector,
    tint: Color, label: String, value: String,
) {
    val gg = LocalGG.current
    Column(
        modifier.ggCard(16.dp).padding(horizontal = 16.dp, vertical = 12.dp),
        verticalArrangement = Arrangement.spacedBy(4.dp),
    ) {
        Row(horizontalArrangement = Arrangement.spacedBy(6.dp), verticalAlignment = Alignment.CenterVertically) {
            Icon(icon, null, tint = tint, modifier = Modifier.size(14.dp))
            Text(label, style = suite(FontWeight.Bold, 12), color = gg.textMuted)
        }
        Text(value, style = suite(FontWeight.ExtraBold, 22), color = gg.text)
    }
}

File: native-ios/Gugu/Features/Session/ResultView.swift
import SwiftUI

// 세션 결과 화면 (ResultScreen.tsx 이식)

struct ResultView: View {
    @Bindable var engine: SessionEngine
    let done: SessionDone
    var onClose: () -> Void

    @Environment(GameStore.self) private var game
    @State private var confetti = 0
    @State private var levelUp = false

    private var result: SessionResult { done.result }
    private var commit: CommitResult { done.commit }
    private var def: ModeDef { Modes.def(result.mode) }

    private var total: Int { result.answers.count }
    private var correct: Int { total - done.wrongCount }
    private var accuracy: Int { total > 0 ? Int((Double(correct) / Double(total) * 100).rounded()) : 0 }
    private var avgMs: Int { total > 0 ? Int((Double(result.answers.reduce(0) { $0 + $1.ms }) / Double(total)).rounded()) : 0 }
    private var best: Int? {
        guard let score = commit.score else { return nil }
        return max(game.state.bestScores[result.mode] ?? 0, score)
    }

    private var headline: String {
        if def.scored { return commit.isNewBest ? "신기록 달성! 🏆" : "수고했어요!" }
        if accuracy >= 95 { return "완벽해요! 🎯" }
        if accuracy >= 70 { return "잘했어요!" }
        return "좋은 시도예요!"
    }

    var body: some View {
        ZStack(alignment: .top) {
            ScrollView {
                VStack(alignment: .leading, spacing: 16) {
                    VStack(alignment: .leading, spacing: 4) {
                        Text("\(def.name) 완료").font(.suite(.bold, 14)).foregroundStyle(Color.gg.textMuted)
                        Text(headline).font(.suite(.extrabold, 28)).foregroundStyle(Color.gg.text)
                    }
                    .padding(.top, 24)

                    if def.scored, let score = commit.score {
                        scoreHero(score)
                    }
                    statGrid
                    if let table = commit.table {
                        masteryRow(table)
                    }
                    if commit.goalReached {
                        Text("🎉 오늘의 목표를 달성했어요!")
                            .font(.suite(.bold, 14)).foregroundStyle(Color.gg.success)
                            .frame(maxWidth: .infinity)
                            .padding(.vertical, 12)
                            .background(Color.gg.success.opacity(0.15), in: RoundedRectangle(cornerRadius: 16, style: .continuous))
                    }

                    Spacer(minLength: 8)
                    actions
                }
                .padding(.horizontal, 24)
                .padding(.bottom, 32)
            }

            ConfettiView(trigger: confetti)
            AchievementToast(ids: commit.unlocked)
                .padding(.horizontal, 20).padding(.top, 12)
            LevelUpOverlay(show: levelUp, level: commit.newLevel) { levelUp = false }
                .animation(.spring(response: 0.4, dampingFraction: 0.7), value: levelUp)
        }
        .frame(maxWidth: .infinity, maxHeight: .infinity)
        .background(Color.gg.bg.ignoresSafeArea())
        .onAppear {
            if accuracy >= 80 || commit.isNewBest { confetti += 1 }
            if commit.leveledUp {
                DispatchQueue.main.asyncAfter(deadline: .now() + 0.6) { levelUp = true }
            }
        }
    }

    private func scoreHero(_ score: Int) -> some View {
        VStack(spacing: 8) {
            if commit.isNewBest {
                Pill(bg: Color.gg.warning.opacity(0.15), fg: .gg.warning) {
                    Image(systemName: "trophy.fill"); Text("신기록!")
                }
            }
            HStack(alignment: .firstTextBaseline, spacing: 4) {
                Text("\(score)").font(.suite(.extrabold, 56)).foregroundStyle(Color.gg.text).monospacedDigit()
                Text("점").font(.suite(.bold, 20)).foregroundStyle(Color.gg.textMuted)
            }
            if let best { Text("최고 기록 \(best)점").font(.suite(.bold, 14)).foregroundStyle(Color.gg.textMuted).monospacedDigit() }
        }
        .frame(maxWidth: .infinity)
        .padding(24)
        .ggCard()
    }

    private var statGrid: some View {
        LazyVGrid(columns: [GridItem(.flexible(), spacing: 12), GridItem(.flexible(), spacing: 12)], spacing: 12) {
            stat(icon: "checkmark", tint: .gg.success, label: "정확도", value: "\(accuracy)%")
            stat(icon: "gauge.medium", tint: .gg.accent, label: "평균 속도", value: String(format: "%.1f초", Double(avgMs) / 1000))
            stat(icon: "sparkles", tint: .gg.warning, label: "획득 XP", value: "+\(commit.xpEarned)")
            stat(icon: "flame.fill", tint: .gg.danger, label: "최고 콤보", value: "\(result.maxCombo)")
        }
    }

    private func stat(icon: String, tint: Color, label: String, value: String) -> some View {
        VStack(alignment: .leading, spacing: 4) {
            HStack(spacing: 6) {
                Image(systemName: icon).font(.system(size: 14, weight: .bold)).foregroundStyle(tint)
                Text(label).font(.suite(.bold, 12)).foregroundStyle(Color.gg.textMuted)
            }
            Text(value).font(.suite(.extrabold, 22)).foregroundStyle(Color.gg.text).monospacedDigit()
        }
        .frame(maxWidth: .infinity, alignment: .leading)
        .padding(.horizontal, 16).padding(.vertical, 12)
        .background(Color.gg.surface, in: RoundedRectangle(cornerRadius: 16, style: .continuous))
        .overlay(RoundedRectangle(cornerRadius: 16, style: .continuous).strokeBorder(Color.gg.border, lineWidth: 1))
    }

    private func masteryRow(_ table: Int) -> some View {
        HStack {
            VStack(alignment: .leading, spacing: 2) {
                Text("\(table)단 마스터리").font(.suite(.bold, 14)).foregroundStyle(Color.gg.text)
                if commit.improvedStars {
                    Text("새 기록 달성!").font(.suite(.bold, 12)).foregroundStyle(Color.gg.accent)
                }
            }
            Spacer()
            StarsView(count: commit.newStars, size: 24)
        }
        .padding(.horizontal, 20).padding(.vertical, 16)
        .ggCard()
    }

    private var actions: some View {
        VStack(spacing: 10) {
            if def.kind == .fixed && done.wrongCount > 0 {
                GGButton(variant: .surface, size: .lg, action: { engine.restart(retryWrong: true) }) {
                    Image(systemName: "arrow.counterclockwise")
                    Text("틀린 문제만 다시 (\(done.wrongCount))")
                }
            }
            GGButton(variant: .primary, size: .lg, action: { engine.restart(retryWrong: false) }) {
                Text("한 판 더")
            }
            GGButton(variant: .ghost, size: .md, action: onClose) {
                Image(systemName: "xmark"); Text("닫기")
            }
        }
    }
}
